Common/Model.cs

```diff
@@ -0,0 +1,143 @@
+
+            // process each mesh located at the current node
+            for (int i = 0; i < node.MeshCount; i++)
+            {
+                // the node object only contains indices to index the actual objects in the scene. 
+                // the scene contains all the data, node is just to keep stuff organized (like relations between nodes).
+                AssimpMesh mesh = scene.Meshes[node.MeshIndices[i]];
+                meshes.Add(ProcessMesh(mesh, transform));
+            }
+
+            for (int i = 0; i < node.ChildCount; i++)
+            {
+                ProcessNode(node.Children[i], scene, transform);
+            }
+        }
+
+        private Mesh ProcessMesh(AssimpMesh mesh, Matrix4x4 transform)
+        {
+            // data to fill
+            List<Vertex> vertices = new List<Vertex>();
+            List<int> indices = new List<int>();
+
+            // walk through each of the mesh's vertices
+            for (int i = 0; i < mesh.VertexCount; i++)
+            {
+                Vertex vertex = new Vertex();
+                // we declare a placeholder vector since assimp uses its own vector class that doesn't directly convert to the Vector3 class so we transfer the data to this placeholder Vector3 first.
+
+                // positions
+                var transformedVertex = transform * mesh.Vertices[i];
+                Vector3 vector;
+                vector.X = transformedVertex.X;
+                vector.Y = transformedVertex.Y;
+                vector.Z = transformedVertex.Z;
+                vertex.Position = vector;
+                // normals
+                if (mesh.HasNormals)
+                {
+                    var transformedNormal = transform * mesh.Normals[i];
+                    vector.X = transformedNormal.X;
+                    vector.Y = transformedNormal.Y;
+                    vector.Z = transformedNormal.Z;
+                    vertex.Normal = vector;
+                }
+                // texture coordinates
+                if (mesh.HasTextureCoords(0)) // does the mesh contain texture coordinates?
+                {
+                    Vector2 vec;
+                    vec.X = mesh.TextureCoordinateChannels[0][i].X;
+                    vec.Y = mesh.TextureCoordinateChannels[0][i].Y;
+                    vertex.TexCoords = vec;
+                }
+                else vertex.TexCoords = new Vector2(0.0f, 0.0f);
+
+                vertices.Add(vertex);
+            }
+            // now walk through each of the mesh's faces (a face is a mesh its triangle) and retrieve the corresponding vertex indices.
+            for (int i = 0; i < mesh.FaceCount; i++)
+            {
+                Face face = mesh.Faces[i];
+                for (int j = 0; j < face.IndexCount; j++)
+                    indices.Add(face.Indices[j]);
+            }
+
+            //Convert to array for fast memory access after inital load
+            return new Mesh(vertices.ToArray(), indices.ToArray());
```

Comment:
We won't be accessing the arrays later so there is nothing fast about doing this copy.
It might be clearer code, but ideally we would upload from the list directly by using the `MemoryMarshal` class.

Reply:
When we create a new Mesh, I would think that we would want the vertices and indices to be stored in arrays since they are sequential in memory (at least they are in C++ according to [learnopengl.com](learnopengl.com)). Each mesh that we process in stored within the model class for drawing using the `Draw()` method. So our final storage type should be as fast as possible. Maybe I have things wrong in my head, would using `MemoryMarshal` be the correct way to do this? Marshaling is still pretty new to me.

Reply:
I've added a comment about how it could be done with `MemoryMarshal` and I've also modified the code in `Mesh` to work with both versions by making it take a `Span<T>`. Probably a little overkill for this demo, but it's good to have samples of this type of code online so people loading in really big meshes don't slow themselves down with loading code that copies the entire mesh multiple times.

Reply:
I certainly appreciate seeing the efficiency that this sample adds.